presto-main/src/test/java/com/facebook/presto/sql/gen/PageProcessorBenchmark.java

```diff
@@ -0,0 +1,211 @@
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.Warmup;
+import org.openjdk.jmh.infra.Blackhole;
+import org.openjdk.jmh.runner.Runner;
+import org.openjdk.jmh.runner.RunnerException;
+import org.openjdk.jmh.runner.options.Options;
+import org.openjdk.jmh.runner.options.OptionsBuilder;
+import org.openjdk.jmh.runner.options.VerboseMode;
+
+import java.util.HashMap;
+import java.util.IdentityHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+import static com.facebook.presto.metadata.FunctionKind.SCALAR;
+import static com.facebook.presto.metadata.MetadataManager.createTestMetadataManager;
+import static com.facebook.presto.operator.scalar.FunctionAssertions.createExpression;
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.spi.type.VarcharType.VARCHAR;
+import static com.facebook.presto.sql.analyzer.ExpressionAnalyzer.getExpressionTypesFromInput;
+import static java.util.Locale.ENGLISH;
+import static java.util.stream.Collectors.toList;
+
+@State(Scope.Thread)
+@OutputTimeUnit(TimeUnit.MILLISECONDS)
+@Fork(10)
+@Warmup(iterations = 10)
+@Measurement(iterations = 10)
+@BenchmarkMode(Mode.AverageTime)
+public class PageProcessorBenchmark
+{
+    private static final Map<String, Type> TYPE_MAP = ImmutableMap.of("bigint", BIGINT, "varchar", VARCHAR);
+    private static final SqlParser SQL_PARSER = new SqlParser();
+    private static final Metadata METADATA = createTestMetadataManager();
+    private static final Session TEST_SESSION = TestingSession.testSessionBuilder().build();
+    private static final int PAGE_COUNT = 10;
+    private static final int POSITIONS = 1024;
+
+    private final Map<Symbol, Type> symbolTypes = new HashMap<>();
+    private final Map<Symbol, Integer> sourceLayout = new HashMap<>();
+
+    private PageProcessor processor;
+    private List<Page> inputPages;
+    private List<? extends Type> types;
+
+    @Param({ "2", "4", "8", "16", "32" })
+    int columnCount;
+
+    @Param({ "varchar", "bigint" })
+    String type;
+
+    @Param({ "false", "true" })
+    boolean dictionaryBlocks;
+
+    @Setup
+    public void setup()
+    {
+        Type type = TYPE_MAP.get(this.type);
+
+        for (int i = 0; i < columnCount; i++) {
+            Symbol symbol = new Symbol(type.getDisplayName().toLowerCase(ENGLISH) + i);
+            symbolTypes.put(symbol, type);
+            sourceLayout.put(symbol, i);
+        }
+
+        List<RowExpression> projections = getProjections(type);
+        types = projections.stream().map(RowExpression::getType).collect(toList());
+
+        inputPages = createPages(PAGE_COUNT, types, dictionaryBlocks);
+        processor = new ExpressionCompiler(createTestMetadataManager()).compilePageProcessor(getFilter(type), projections);
+    }
+
+    @Benchmark
+    public void rowOriented(Blackhole b)
+    {
+        for (int i = 0; i < PAGE_COUNT; i++) {
```

Comment:
It's generally not a good idea to loop in the benchmark methods. This can result in measurement errors due to loop unrolling and other optimizations. See http://hg.openjdk.java.net/code-tools/jmh/file/bcec9a03787f/jmh-samples/src/main/java/org/openjdk/jmh/samples/JMHSample_11_Loops.java for more details